health/health.go

```diff
@@ -0,0 +1,207 @@
+package health
+
+import (
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var (
+	mutex            sync.RWMutex
+	registeredChecks = make(map[string]Checker)
+)
+
+// Checker is the interface for a Health Checker
+type Checker interface {
+	// Check returns nil if the service is okay.
+	Check() error
+}
+
+// CheckFunc is a convenience type to create functions that implement
+// the Checker interface
+type CheckFunc func() error
+
+// Check Implements the Checker interface to allow for any func() error method
+// to be passed as a Checker
+func (cf CheckFunc) Check() error {
+	return cf()
+}
+
+// Updater implements a health check that is explicitly set.
+type Updater interface {
+	Checker
+
+	// Update updates the current status of the health check.
+	Update(status error)
+}
+
+// updater implements Checker and Updater, providing an asynchronous Update
+// method.
+// This allows us to have a Checker that returns the Check() call immediately
+// not blocking on a potentially expensive check.
+type updater struct {
+	mu     sync.Mutex
+	status error
+}
+
+// Check implements the Checker interface
+func (u *updater) Check() error {
+	u.mu.Lock()
+	defer u.mu.Unlock()
```

Comment:
@icecrime Is this based on style or perceived performance? Typically, its idiomatic to do locking in a trivial function, like this, then have an inner function to perform more complex behavior:

```
func (u *Updater) Check() error {
	u.mu.Lock()
	defer u.mu.Unlock()

        return u.check()
}

func (u *Updater) check() error {
	// actual check implementation
}
```

If this advice is based on performance, we should only diverge from Go style when it can be proven. If it's based on style, we need to have another discussion, as defer should be considered good style for resource cleanup.